server/src/main/java/org/elasticsearch/cluster/metadata/SystemIndexMetadataUpgradeService.java

```diff
@@ -36,41 +37,55 @@ public class SystemIndexMetadataUpgradeService implements ClusterStateListener {
     private final SystemIndices systemIndices;
     private final ClusterService clusterService;
 
-    private boolean master = false;
-
-    private volatile Map<String, IndexMetadata> lastIndexMetadataMap = ImmutableOpenMap.of();
     private volatile boolean updateTaskPending = false;
 
+    private volatile long triggeredVersion = -1L;
+
     public SystemIndexMetadataUpgradeService(SystemIndices systemIndices, ClusterService clusterService) {
         this.systemIndices = systemIndices;
         this.clusterService = clusterService;
     }
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        if (event.localNodeMaster() != master) {
-            this.master = event.localNodeMaster();
-        }
-
-        if (master && updateTaskPending == false) {
+        if (updateTaskPending == false
+            && event.localNodeMaster()
+            && (event.previousState().nodes().isLocalNodeElectedMaster() == false
+                || event.state().metadata().indices() != event.previousState().metadata().indices())) {
             final Map<String, IndexMetadata> indexMetadataMap = event.state().metadata().indices();
-
-            if (lastIndexMetadataMap != indexMetadataMap) {
-                for (Map.Entry<String, IndexMetadata> cursor : indexMetadataMap.entrySet()) {
-                    if (cursor.getValue() != lastIndexMetadataMap.get(cursor.getKey())) {
-                        IndexMetadata indexMetadata = cursor.getValue();
-                        boolean requiresUpdate = requiresUpdate(indexMetadata);
-                        if (requiresUpdate) {
-                            updateTaskPending = true;
-                            submitUnbatchedTask(
-                                "system_index_metadata_upgrade_service {system metadata change}",
-                                new SystemIndexMetadataUpdateTask()
-                            );
-                            break;
+            final var previousIndices = event.previousState().metadata().indices();
+            final long triggerV = event.state().version();
+            triggeredVersion = triggerV;
+            // Fork to the management pool to avoid blocking the cluster applier thread unnecessarily for very large index counts
+            // TODO: we should have a more efficient way of getting just the changed indices so that we don't have to fork here
+            clusterService.threadPool().executor(ThreadPool.Names.MANAGEMENT).execute(new AbstractRunnable() {
+                @Override
+                protected void doRun() {
+                    if (triggeredVersion != triggerV) {
```

Comment:
I just want to make sure I understand what is happening here. Example scenario:

1. Update A happens, and index metadata is changed. We set the volatile `triggeredVersion` field to this change event's version, and submit a new task to the management thread pool.
2. Update B happens, and index metadata is changed again. We set the volatile `triggeredVersion` field to B's version, and submit a new task to the management thread pool.
3. The management thread pool begins processing the task from update A, but `triggeredVersion` has changed, so we return without running expensive updates.
4. The management thread pool begins processing the task from update B, and finds that `triggeredVersion` has not changed, so we submit our tasks to update system index metadata.

Since updating index metadata is really a matter of taking current index metadata and making it consistent with system expectations, we don't need to worry about dropping these intermediate updates. As long as we have processed the most recent update, our metadata should end up in a good state.

Am I getting anything wrong?

Reply:
No, you're spot on :)